pandera/dtypes_.py

```diff
@@ -0,0 +1,286 @@
+import functools
+from dataclasses import dataclass, field
+from typing import Any, Tuple, Type, Union
+
+try:  # python 3.8+
+    from typing import Literal  # type: ignore
+except ImportError:
+    from typing_extensions import Literal  # type: ignore
+
+
+def immutable(dtype=None, **kwargs) -> Type:
+    dataclass_kwargs = {"frozen": True, "init": False, "repr": False}
+    dataclass_kwargs.update(kwargs)
+
+    if dtype is None:
+        return functools.partial(dataclass, **dataclass_kwargs)
+    return dataclass(**dataclass_kwargs)(dtype)
+
+
+class DisableInitMixin:
```

Comment:
what's the reasoning behind this mixin?

I'm guessing the `dataclass(init=...)` doesn't have an effect here since `DataType` already defines an `__init__` method.

Another thought is to consider wrapping this functionality into `immutable`... I think the fewer constructs users need to worry about the better (in this context, user means "advanced user who wants to extend the type system, including us" 😃 .

Reply:
I wanted `numpy_engine.DataType` and `pandas_engine.DataType` to have an `__init__(self, dtype:Any)` to fall back to if the engine doesn't recognize the dtype: https://github.com/jeffzi/pandera/blob/3cd08e85be75449a2b1a13b3245bb112d5d5d749/pandera/engines/pandas_engine.py#L78-L81. 

That could happen if the user provides a custom extension dtype. In that case, the default coerce could work. Maybe we need a warning if DataType is instantiated by the engine.

> I'm guessing the dataclass(init=...) doesn't have an effect here since DataType already defines an __init__ method.

Correct !

> Another thought is to consider wrapping this functionality into immutable.

Agreed and done !

I had to do multiple things:
 1. `immutable` adds an empty  `__init__` if `init=False`
 2. Disable init from `pandas/numpy_engine.DataType.type`: `type: Any = dataclasses.field(repr=False, init=False)`. The 
     point is that sub-classes will never have `type` added to a generated init.
 3. Add a custom `pandas/numpy_engine.DataType.__init__(self, dtype: Any)`
 4. Decorate `pandas/numpy_engine.DataType` with `@immutable(init=True)` so that my custom `__init__ ` is not 
    overriden.
 
 It's convoluted but works :sweat_smile: